### ISSUE TITLE
Cannot start machine with podman version 4.1.0 in M1 Mac

### ISSUE THREAD

**Issue by Charizard** [containers/podman#14303]
/kind bug

**Description**

In a M1 Mac, Downloaded the VM via `podman machine init` and then when trying to start it via `podman machine start` I get thew following error: 
```
Starting machine "podman-machine-default"
Waiting for VM ...
Error: dial unix /var/folders/kk/vvrx0k6x2b54wnnhc0zt6rv00000gr/T/podman/podman-machine-default_ready.sock: connect: connection refused
```
Tried to download the arch specific VM via `podman machine init --image-path=https://builds.coreos.fedoraproject.org/prod/streams/stable/builds/35.20220424.3.0/aarch64/fedora-coreos-35.20220424.3.0-qemu.aarch64.qcow2.xz` and starting it again with the same error.

**Steps to reproduce the issue:**

1. `podman machine init`

2. `podman machine start`

**Describe the results you received:**

Error when trying to start the machine.

**Describe the results you expected:**

Expected the machine to start successfully.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
Cannot connect to Podman. Please verify your connection to the Linux system using `podman system connection list`, or try `podman machine init` and `podman machine start` to manage a new Linux VM
Error: unable to connect to Podman. failed to create sshClient: connection to bastion host (ssh://core@localhost:56180/run/user/504/podman/podman.sock) failed: dial tcp [::1]:56180: connect: connection refused
```

**Output of `podman info --debug`:**

```
Cannot connect to Podman. Please verify your connection to the Linux system using `podman system connection list`, or try `podman machine init` and `podman machine start` to manage a new Linux VM
Error: unable to connect to Podman. failed to create sshClient: connection to bastion host (ssh://core@localhost:56180/run/user/504/podman/podman.sock) failed: dial tcp [::1]:56180: connect: connection refused
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman: stable 4.1.0 (bottled), HEAD
Tool for managing OCI containers and pods
https://podman.io/
/usr/local/Cellar/podman/4.1.0 (174 files, 48.2MB) *
  Poured from bottle on 2022-05-19 at 17:51:55
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/podman.rb
License: Apache-2.0
==> Dependencies
Build: go ✘, go-md2man ✘
Required: qemu ✔
==> Options
--HEAD
        Install HEAD version
==> Caveats
zsh completions have been installed to:
  /usr/local/share/zsh/site-functions
==> Analytics
install: 20,290 (30 days), 55,705 (90 days), 143,980 (365 days)
install-on-request: 20,245 (30 days), 55,655 (90 days), 143,917 (365 days)
build-error: 0 (30 days)
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide? (https://github.com/containers/podman/blob/main/troubleshooting.md)**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**

**Comment by phenri00** [comment]
Experiencing similar(?) issue. M1 Mac.

```
a@M1-Air ~ $ podman machine start
Starting machine "podman-machine-default"
Waiting for VM ...
Error: dial unix /var/folders/1d/ts5cfq0d5sqfx5h9k7vkh2000000gn/T/podman/podman-machine-default_ready.sock: connect: connection refused

a@M1-Air ~ $ podman info --debug
Cannot connect to Podman....
```

```
a@M1-Air ~ $ brew info podman
podman: stable 4.1.0 (bottled), HEAD
Tool for managing OCI containers and pods
https://podman.io/
/usr/local/Cellar/podman/HEAD-948c5e9 (174 files, 47.8MB) *
  Built from source on 2022-05-19 at 14:15:13
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/podman.rb
License: Apache-2.0
==> Dependencies
Build: go ✔, go-md2man ✔
Required: qemu ✔
==> Options
--HEAD
	Install HEAD version
```

**Comment by nycnewman** [comment]
Getting same issue. If I run in debug mode the following command is run which produces the error if run standalone:

/opt/homebrew/bin/qemu-system-aarch64 -m 27000 -smp 10 -fw_cfg name=opt/com.coreos/config,file=/Users/edwardnewman/.config/containers/podman/machine/qemu/podman-machine-default.ign -qmp unix://var/folders/tz/4n8rgr6n6wj5cgqzjx1dh0_80000gq/T/podman/qmp_podman-machine-default.sock,server=on,wait=off -netdev socket,id=vlan,fd=3 -device virtio-net-pci,netdev=vlan,mac=5a:94:ef:e4:0c:ee -device virtio-serial -chardev socket,path=/var/folders/tz/4n8rgr6n6wj5cgqzjx1dh0_80000gq/T/podman/podman-machine-default_ready.sock,server=on,wait=off,id=podman-machine-default_ready -device virtserialport,chardev=podman-machine-default_ready,name=org.fedoraproject.port.0 -accel hvf -accel tcg -cpu cortex-a57 -M virt,highmem=off -drive file=/opt/homebrew/share/qemu/edk2-aarch64-code.fd,if=pflash,format=raw,readonly=on -drive file=/Users/edwardnewman/.local/share/containers/podman/machine/qemu/podman-machine-default_ovmf_vars.fd,if=pflash,format=raw -virtfs local,path=/Users/edwardnewman,mount_tag=vol0,security_model=mapped-xattr -drive if=virtio,file=/Users/edwardnewman/.local/share/containers/podman/machine/qemu/podman-machine-default_fedora-coreos-36.20220511.dev.0-qemu.aarch64.qcow2
**qemu-system-aarch64: -netdev socket,id=vlan,fd=3: can't get socket option SO_TYPE**

**Comment by nycnewman** [comment]
Note that it is not possible on MacOS to downgrade to previous podman version as the tap appears to only understand 4.1.0. Was proposing to test to see if that still works as issue appears to be with qemu.

**Comment by mheon** [comment]
@baude @ashley-cui PTAL

**Comment by ktram** [comment]
Also running into this, and I saw the same `qemu-system-aarch64: -netdev socket,id=vlan,fd=3: can't get socket option SO_TYPE` error message as @nycnewman.  After downgrading qemu, I was able to run podman.

**Comment by lukas-vlcek** [comment]
I am running into the same issue.
@ktram Would you mind sharing your steps to downgrade qemu? I installed it using brew and it seems that downgrading packages in brew is full of magic.

**Comment by ashley-cui** [comment]
Hmm, maybe it looks like something changed in qemu 7.0? Taking a look...

**Comment by shanesmith** [comment]
In case it's of any use, I'm finding that the max memory I'm able to set on a machine is `3072`, otherwise I see the `connection refused` error.

✔️ The following works:
```
podman machine init # default memory is 2048
podman machine start
```

```
podman machine init -m 3072
podman machine start
```

```
podman machine init
podman machine set -m 3072
podman machine start
```


✖️ The following results in the `connection refused` error:

```
podman machine init -m 3073
podman machine start
```

```
podman machine init -m 4096
podman machine start
```

```
podman machine init
podman machine set -m 4096
podman machine start
```

**Comment by justin-f-perez** [comment]
I am also experiencing this issue on an M1 mac running MacOS 12.4 (21F79) with podman 4.1.0 and qemu 7.0.0 (with qemu being installed automatically by homebrew as a dependency of podman). 

## 🔧  pm-clean util function
I ran this before each of the below tests
```bash
pm-clean() { podman machine ls -q | sed 's/\*$//' | xargs -L1 podman machine rm --force 2>/dev/null; podman machine ls --noheading; }
```

## ❌ 🐛  `podman machine init; podman machine start podman-machine-default`

**For whatever reason, this spawns a separate qemu-system-aarch64 window now rather than running in the background as it's supposed to.**

## ✅  `podman machine init --cpus=4 --disk-size=20 --memory=3072 --timezone=local --rootful=false --now --log-level=trace`

Confirming @shanesmith comment, rootless works with memory <= 3072

## ✅  `podman machine init --cpus=4 --disk-size=20 --memory=3072 --timezone=local --rootful=true --now --log-level=trace`

So does rootful (but only with memory <= 3072)

## ❌ `podman machine init --cpus=4 --disk-size=20 --memory=4096 --timezone=local --rootful=true --now --log-level=trace`
```
time="2022-05-25T15:12:10-05:00" level=info msg="podman filtering at log level trace"
Extracting compressed file
time="2022-05-25T15:12:28-05:00" level=trace msg="Reading configuration file \"/Users/podman-rootful-user/.config/containers/containers.conf\""
time="2022-05-25T15:12:28-05:00" level=trace msg="Reading configuration file \"/Users/podman-rootful-user/.config/containers/containers.conf\""
Image resized.
Machine init complete
time="2022-05-25T15:12:28-05:00" level=debug msg="[/opt/homebrew/bin/qemu-system-aarch64 -m 4096 -smp 4 -fw_cfg name=opt/com.coreos/config,file=/Users/podman-rootful-user/.config/containers/podman/machine/qemu/podman-machine-default.ign -qmp unix://var/folders/8_/_z_4l9jx6214q4ztktdvhg3d19bx8k/T/podman/qmp_podman-machine-default.sock,server=on,wait=off -netdev socket,id=vlan,fd=3 -device virtio-net-pci,netdev=vlan,mac=5a:94:ef:e4:0c:ee -device virtio-serial -chardev socket,path=/var/folders/8_/_z_4l9jx6214q4ztktdvhg3d19bx8k/T/podman/podman-machine-default_ready.sock,server=on,wait=off,id=podman-machine-default_ready -device virtserialport,chardev=podman-machine-default_ready,name=org.fedoraproject.port.0 -accel hvf -accel tcg -cpu cortex-a57 -M virt,highmem=off -drive file=/opt/homebrew/share/qemu/edk2-aarch64-code.fd,if=pflash,format=raw,readonly=on -drive file=/Users/podman-rootful-user/.local/share/containers/podman/machine/qemu/podman-machine-default_ovmf_vars.fd,if=pflash,format=raw -virtfs local,path=/Users/podman-rootful-user,mount_tag=vol0,security_model=mapped-xattr -drive if=virtio,file=/Users/podman-rootful-user/.local/share/containers/podman/machine/qemu/podman-machine-default_fedora-coreos-36.20220511.dev.0-qemu.aarch64.qcow2]"
Waiting for VM ...
Error: dial unix /var/folders/8_/_z_4l9jx6214q4ztktdvhg3d19bx8k/T/podman/podman-machine-default_ready.sock: connect: connection refused
```

## ❌ `podman machine init --cpus=4 --disk-size=20 --memory=4096 --timezone=local --rootful=false --now --log-level=trace`
```
time="2022-05-25T15:01:24-05:00" level=info msg="podman filtering at log level trace"
Extracting compressed file
time="2022-05-25T15:01:42-05:00" level=trace msg="Reading configuration file \"/Users/podman-rootless-user/.config/containers/containers.conf\""
time="2022-05-25T15:01:42-05:00" level=trace msg="Reading configuration file \"/Users/podman-rootless-user/.config/containers/containers.conf\""
Image resized.
Machine init complete
time="2022-05-25T15:01:42-05:00" level=debug msg="[/opt/homebrew/bin/qemu-system-aarch64 -m 4096 -smp 4 -fw_cfg name=opt/com.coreos/config,file=/Users/podman-rootless-user/.config/containers/podman/machine/qemu/podman-machine-default.ign -qmp unix://var/folders/8_/_z_4l9jx6214q4ztktdvhg3d19bx8k/T/podman/qmp_podman-machine-default.sock,server=on,wait=off -netdev socket,id=vlan,fd=3 -device virtio-net-pci,netdev=vlan,mac=5a:94:ef:e4:0c:ee -device virtio-serial -chardev socket,path=/var/folders/8_/_z_4l9jx6214q4ztktdvhg3d19bx8k/T/podman/podman-machine-default_ready.sock,server=on,wait=off,id=podman-machine-default_ready -device virtserialport,chardev=podman-machine-default_ready,name=org.fedoraproject.port.0 -accel hvf -accel tcg -cpu cortex-a57 -M virt,highmem=off -drive file=/opt/homebrew/share/qemu/edk2-aarch64-code.fd,if=pflash,format=raw,readonly=on -drive file=/Users/podman-rootless-user/.local/share/containers/podman/machine/qemu/podman-machine-default_ovmf_vars.fd,if=pflash,format=raw -virtfs local,path=/Users/podman-rootless-user,mount_tag=vol0,security_model=mapped-xattr -drive if=virtio,file=/Users/podman-rootless-user/.local/share/containers/podman/machine/qemu/podman-machine-default_fedora-coreos-36.20220511.dev.0-qemu.aarch64.qcow2]"
Waiting for VM ...
Error: dial unix /var/folders/8_/_z_4l9jx6214q4ztktdvhg3d19bx8k/T/podman/podman-machine-default_ready.sock: connect: connection refused
```

**Comment by justin-f-perez** [comment]
@shanesmith It doesn't seem to matter how I initialize the machine, I get the same meaningless error

```bash
$ pm machine stop && pm machine rm --force && pm machine init && pm machine start --log
INFO[0000] podman filtering at log level trace
DEBU[0000] [/opt/homebrew/bin/qemu-system-aarch64 -m 2048 -smp 1 -fw_cfg name=opt/com.coreos/config,file=/Users/podman-is-so-broken/.config/containers/podman/machine/qemu/podman-machine-default.ign -qmp unix://var/folders/8_/_z_4l9jx6214q4ztktdvhg3d19bx8k/T/podman/qmp_podman-machine-default.sock,server=on,wait=off -netdev socket,id=vlan,fd=3 -device virtio-net-pci,netdev=vlan,mac=5a:94:ef:e4:0c:ee -device virtio-serial -chardev socket,path=/var/folders/8_/_z_4l9jx6214q4ztktdvhg3d19bx8k/T/podman/podman-machine-default_ready.sock,server=on,wait=off,id=podman-machine-default_ready -device virtserialport,chardev=podman-machine-default_ready,name=org.fedoraproject.port.0 -accel hvf -accel tcg -cpu cortex-a57 -M virt,highmem=off -drive file=/opt/homebrew/share/qemu/edk2-aarch64-code.fd,if=pflash,format=raw,readonly=on -drive file=/Users/podman-is-so-broken/.local/share/containers/podman/machine/qemu/podman-machine-default_ovmf_vars.fd,if=pflash,format=raw -virtfs local,path=/Users/podman-is-so-broken,mount_tag=vol0,security_model=mapped-xattr -drive if=virtio,file=/Users/podman-is-so-broken/.local/share/containers/podman/machine/qemu/podman-machine-default_fedora-coreos-36.20220511.dev.0-qemu.aarch64.qcow2]
Starting machine "podman-machine-default"
DEBU[0000] [/opt/homebrew/bin/qemu-system-aarch64 -m 2048 -smp 1 -fw_cfg name=opt/com.coreos/config,file=/Users/podman-is-so-broken/.config/containers/podman/machine/qemu/podman-machine-default.ign -qmp unix://var/folders/8_/_z_4l9jx6214q4ztktdvhg3d19bx8k/T/podman/qmp_podman-machine-default.sock,server=on,wait=off -netdev socket,id=vlan,fd=3 -device virtio-net-pci,netdev=vlan,mac=5a:94:ef:e4:0c:ee -device virtio-serial -chardev socket,path=/var/folders/8_/_z_4l9jx6214q4ztktdvhg3d19bx8k/T/podman/podman-machine-default_ready.sock,server=on,wait=off,id=podman-machine-default_ready -device virtserialport,chardev=podman-machine-default_ready,name=org.fedoraproject.port.0 -accel hvf -accel tcg -cpu cortex-a57 -M virt,highmem=off -drive file=/opt/homebrew/share/qemu/edk2-aarch64-code.fd,if=pflash,format=raw,readonly=on -drive file=/Users/podman-is-so-broken/.local/share/containers/podman/machine/qemu/podman-machine-default_ovmf_vars.fd,if=pflash,format=raw -virtfs local,path=/Users/podman-is-so-broken,mount_tag=vol0,security_model=mapped-xattr -drive if=virtio,file=/Users/podman-is-so-broken/.local/share/containers/podman/machine/qemu/podman-machine-default_fedora-coreos-36.20220511.dev.0-qemu.aarch64.qcow2]
Waiting for VM ...
Mounting volume... /Users/podman-is-so-broken:/Users/podman-is-so-broken
DEBU[0013] Executing: ssh [-i /Users/podman-is-so-broken/.ssh/podman-machine-default -p 62930 core@localhost -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no -q -- sudo chattr -i / ; sudo mkdir -p /Users/podman-is-so-broken ; sudo chattr +i / ;]
Error: exit status 255

$ echo $?
125
```

**Comment by ktram** [comment]
@justin-f-perez I don't know if this will help you, but I used the information at the bottom of [this](https://itnext.io/goodbye-docker-desktop-hello-minikube-3649f2a1c469) to configure podman and minikube on my M1.  I've been initializing with

```sh
podman machine init --cpus 6 --memory 12288 --disk-size 50 --image-path next
```

**Comment by justin-f-perez** [comment]
@ashley-cui what command should be run to see what VM image is being used?

@ktram many thanks for the suggestion, but this causes the same error as described in the initial issue description. I also tried just using just `--image-path next`. Finally, since the error occurs right after the message about mounting, I attempted to tell `podman machine init` not to mount any volumes by passing `-v ':'`, which led to the same error as the initial issue description

## ❌  `podman machine stop && podman machine rm --force && podman machine init --memory 12288 --cpus 6 --disk-size=50 --image-path next && podman machine start`
```Machine "podman-machine-default" stopped successfully
Extracting compressed file
Image resized.
Machine init complete
To start your machine run:

	podman machine start

Starting machine "podman-machine-default"
Waiting for VM ...
Error: dial unix /var/folders/8_/_z_4l9jx6214q4ztktdvhg3d19bx8k/T/podman/podman-machine-default_ready.sock: connect: connection refused
```

## ❌ `podman machine stop && podman machine rm --force && podman machine init --image-path next && podman machine start`
```
Machine "podman-machine-default" stopped successfully
Extracting compressed file
Image resized.
Machine init complete
To start your machine run:

	podman machine start

Starting machine "podman-machine-default"
Waiting for VM ...
Mounting volume... /Users/jfperez:/Users/jfperez
Error: exit status 255
```

## ❌  `podman machine stop && podman machine rm --force && podman machine init --image-path next -v ':' && podman machine start`
```
Machine "podman-machine-default" stopped successfully
Extracting compressed file
Image resized.
Machine init complete
To start your machine run:

	podman machine start

Starting machine "podman-machine-default"
Waiting for VM ...
Error: dial unix /var/folders/8_/_z_4l9jx6214q4ztktdvhg3d19bx8k/T/podman/podman-machine-default_ready.sock: connect: connection refused
```

**Comment by ashley-cui** [comment]
@justin-f-perez `podman machine inspect`

**Comment by justin-f-perez** [comment]
@ashley-cui 
```json
[
     {
          "ConfigPath": {
               "Path": "/Users/me/.config/containers/podman/machine/qemu/podman-machine-default.json"
          },
          "Created": "2022-05-26T15:06:54.843899-05:00",
          "Image": {
               "IgnitionFilePath": {
                    "Path": "/Users/me/.config/containers/podman/machine/qemu/podman-machine-default.ign"
               },
               "ImageStream": "next",
               "ImagePath": {
                    "Path": "/Users/me/.local/share/containers/podman/machine/qemu/podman-machine-default_fedora-coreos-36.20220511.dev.0-qemu.aarch64.qcow2"
               }
          },
          "LastUp": "0001-01-01T00:00:00Z",
          "Name": "podman-machine-default",
          "Resources": {
               "CPUs": 1,
               "DiskSize": 100,
               "Memory": 2048
          },
          "SSHConfig": {
               "IdentityPath": "/Users/me/.ssh/podman-machine-default",
               "Port": 51642,
               "RemoteUsername": "core"
          },
          "State": ""
     }
]
```

**Comment by raycardillo** [comment]
Just ran into this myself as well. On an M1 Macbook Pro. Ended up being the memory limit for me. All I did that caused this was:
- typical brew install (4.1.0) + typical default podman machine init/start
- realized I needed more memory so I `rm` the initial machine
- created a new machine with `-m 4096`
- pain, pain, pain (`Cannot connect to Podman - failed to create sshClient`)
- found this issue and tried to recreate the machine with `-m 3072`
- back up and running (but not with the memory config I wanted)

**Comment by cpolizzi** [comment]
I can also confirm that I have the same issue.

**Comment by justin-f-perez** [comment]
I figured out how to downgrade to 4.0.3 (which is working as expected) and wrote a little script to help others

```shell
#!/bin/zsh
# shellcheck shell=bash
#
# This script creates a homebrew tap for an old version of podman.
# If no version is provided as the first argument, a list of
# versions is either presented for selection via fzf (if installed)
# or printed to stdout.
set -euo pipefail

list_versions() {
  cd "$(brew tap-info homebrew/core | grep '^/' | cut -d' ' -f1)"
  git log -p -G 'url.*podman' -- 'Formula/podman.rb' |
    grep -e "-.*url.*https://github.com/containers/podman/archive/v.*.tar.gz" | sed 's/^.*v\([0-9.]*\)\.tar.gz.*/\1/'
}

create_tap() {
  version="$1"
  brew tap-new "$USER/local-podman" ||:
  brew extract --version="${version}" podman "$USER/local-podman"
}

version="${1:-}"
if [[ -z "${version}" ]]; then
  if command -v fzf; then
    echo "Listing versions (this may take a minute)" && sleep 2
    version="$(list_versions | fzf)"
  else
    echo "Usage: ./poldman.zsh <VERSION>"
    echo "Example:"
    echo "./poldman.zsh 4.0.3"
    echo "Listing versions (this may take a minute)"
    list_versions
  fi
fi


podman machine stop ||:
create_tap "${version}"
brew install "podman@${version}"
brew link "podman@${version}"
```

Hope this helps!

**Comment by arixmkii** [comment]
Had the same issue with Qemu from homebrew. Downgraded to 6.2.0_1 and everything is ok. My workaround here https://github.com/Homebrew/homebrew-core/issues/102242#issuecomment-1137108389

**Comment by shanesmith** [comment]
Potentially related: https://gitlab.com/qemu-project/qemu/-/issues/903

**Comment by justin-f-perez** [comment]
I realized today that downgrading to `podman@4.0.3` resolved the other problems I seem to be having that others aren't, but the memory limit issue persists. Notably, I get the following error with Qemu 7.0 and podman 4.0.3:
```
qemu-system-aarch64: Addressing limited to 32 bits, but memory exceeds it by <NUMBER> bytes

Error: dial unix ... connect: connection refused
```

It's not clear why podman 4.1.0 seems to supress the `Addressing limited to...` error message even when log level is set to maximum verbosity

**Comment by shanesmith** [comment]
I'm able to start the machine with memory greater than 3072 with the following changes to `CmdLine` in `~/.config/containers/podman/machine/qemu/<name>.json`

- Set `-cpu` to `host`
- Set `-M` to `virt` (ie: remove `highmem=off`)

Tried a handful of other cpus listed by `qemu-system-aarch64 -cpu help` and of those I tried  `max` also works.

I really don't know what the implications of these changes are though, any `qemu` savants in the house? =)

**Comment by jorneilander** [comment]
> I'm able to start the machine with memory greater than 3072 with the following changes to `CmdLine` in `~/.config/containers/podman/machine/qemu/<name>.json`
> 
>     * Set `-cpu` to `host`
> 
>     * Set `-M` to `virt` (ie: remove `highmem=off`)

Doing this magically resets my Mac 🆘 

```
QEMU emulator version 7.0.0
podman 4.1.0
OS X 12.3.1
```

**Comment by Va1** [comment]
@shanesmith legend!

editing `~/.config/containers/podman/machine/qemu/<podman machine name>.json` upon machine creation, in the way described above, made things functional for me again on Max OS 12.4 with any desired amount of memory.

hope this gets addressed soon in a way which would make things work out of the box for everyone.

**Comment by sjwl** [comment]
My workaround is to delete the VM and start from scratch
```
$ podman machine rm --force
$ podman machine init --cpus 2 --memory 2048 --disk-size 20
$ podman machine start
```
version
```
$ podman --version
podman version 4.1.0
$ qemu-system-aarch64 --version 
QEMU emulator version 7.0.0
```

**Comment by Nilegfx** [comment]
downgrading helped in my case 

``` 
  brew uninstall --ignore-dependencies qemu
```

then

```
curl -L -H "Authorization: Bearer QQ==" -o qemu-6.2.0_1.monterey.bottle.tar.gz https://ghcr.io/v2/homebrew/core/qemu/blobs/sha256:fcc3b1a8139f70dae57f5449f3856f9b3b67448ee0623e64da1e47dc255b46f6
brew install -f qemu-6.2.0_1.monterey.bottle.tar.gz
```

![image](https://user-images.githubusercontent.com/1198975/172680517-234c2b87-e63b-4b45-9979-4d8812dcb541.png)

**Comment by cpolizzi** [comment]
CPU setting of `cortex-a57` is wrong. It should be `host`. Memory model should be `himem=on` and not `highmem=off`.

Podman managed machine JSON configuration file needs to be this:
````
 "-cpu",
  "host",
  "-M",
  "virt,highmem=on",
```

Instead of this:
```
  "-cpu",
  "cortex-a57",
  "-M",
  "virt,highmem=off",
```

Sorry, but it's not QEMU that is the problem. It is the way that Podman is instructing QEMU what both the CPU architecture model and memory availability model looks like. Both of which are incorrect.

**Comment by e-minguez** [comment]
> CPU setting of `cortex-a57` is wrong. It should be `host`. Memory model should be `himem=on` and not `highmem=off`.
> 
> Podman managed machine JSON configuration file needs to be this:
> 
> ```
>  "-cpu",
>   "host",
>   "-M",
>   "virt,highmem=on",
> ```
> 
> Instead of this:
> 
> ```
>   "-cpu",
>   "cortex-a57",
>   "-M",
>   "virt,highmem=off",
> ```
> 
> Sorry, but it's not QEMU that is the problem. It is the way that Podman is instructing QEMU what both the CPU architecture model and memory availability model looks like. Both of which are incorrect.

Does this mean you can edit the `~/.config/containers/podman/machine/qemu/podman-machine-default.json` file to add those changes and it would work?

Thanks.

**Comment by ashley-cui** [comment]
Thanks @cpolizzi @shanesmith! I'll open up a fix for this now.

**Comment by cpolizzi** [comment]
> Does this mean you can edit the `~/.config/containers/podman/machine/qemu/podman-machine-default.json` file to add those changes and it would work?
> 
> Thanks.

Yes. I do it all of the time without issue for various reasons.

**Comment by shanesmith** [comment]
Works fine on macOS 12.4, but these qemu flag changes cause a kernel panic and crashes macOS when on version 12.3.

@cpolizzi since you seem to know a thing or two about qemu, would you happen to know why?

**Comment by cpolizzi** [comment]
@shanesmith I did not get a kernel panic on MacOS 12.3.1:
```
$ sw_vers 
ProductName:	macOS
ProductVersion:	12.3.1
BuildVersion:	21E258

$ uname -m
arm64
```



# Podman 3.4.4 Testing

Test on Podman 3.4.4, due to my customer needs currently, **fedora-coreos-35.20220424.3.0-qemu.aarch64.qcow2.xz**:
```
$ brew tap-new local/podman
$ brew extract --version 3.4.4 podman local/podman
$ brew install podman@3.4.4

$ brew list --versions qemu
qemu 7.0.0

$ podman --version
podman version 3.4.4

$ podman machine init \
    --memory 6144 \
    --cpus 4 \
    --disk-size 40 \
    --image-path https://builds.coreos.fedoraproject.org/prod/streams/stable/builds/35.20220424.3.0/aarch64/fedora-coreos-35.20220424.3.0-qemu.aarch64.qcow2.xz
Downloading VM image: fedora-coreos-35.20220424.3.0-qemu.aarch64.qcow2.xz: done  
Extracting compressed file
Image resized.

$ podman machine list
NAME                     VM TYPE     CREATED        LAST UP        CPUS        MEMORY      DISK SIZE
podman-machine-default*  qemu        7 minutes ago  7 minutes ago  4           6.442GB     42.95GB

$ podman machine start
INFO[0000] waiting for clients...                       
INFO[0000] listening tcp://127.0.0.1:7777               
INFO[0000] new connection from  to /var/folders/77/qk6_53m11yb7t885qf6d32jw0000gn/T/podman/qemu_podman-machine-default.sock 
Waiting for VM ...
qemu-system-aarch64: Addressing limited to 32 bits, but memory exceeds it by 3221225472 bytes

Error: dial unix /var/folders/77/qk6_53m11yb7t885qf6d32jw0000gn/T/podman/podman-machine-default_ready.sock: connect: connection refused
ERRO[0003] cannot receive packets from , disconnecting: cannot read size from socket: EOF 
ERRO[0003] cannot read size from socket: EOF
```

Making the changes to **~/.config/containers/podman/machine/qemu/podman-machine-default.json** QEMU JSON configuration for CPU and memory model as suggested results in succes. Diff of the JSON file:
```
28c28
<   "cortex-a57",
---
>   "host",
30c30
<   "virt,highmem=off",
---
>   "virt,highmem=on",
51c51
< }
\ No newline at end of file
---
> }
```

Then starting the VM again results in:
```
$ podman machine start
podman machine start                                                   
INFO[0000] waiting for clients...                       
INFO[0000] listening tcp://127.0.0.1:7777               
INFO[0000] new connection from  to /var/folders/77/qk6_53m11yb7t885qf6d32jw0000gn/T/podman/qemu_podman-machine-default.sock 
Waiting for VM ...
Machine "podman-machine-default" started successfully

$ podman version
Client:
Version:      3.4.4
API Version:  3.4.4
Go Version:   go1.18.3
Built:        Wed Dec  8 12:41:11 2021
OS/Arch:      darwin/arm64

Server:
Version:      3.4.4
API Version:  3.4.4
Go Version:   go1.16.8
Built:        Wed Dec  8 15:48:10 2021
OS/Arch:      linux/arm64

$ podman machine ssh cat /etc/os-release | grep ^VERSION=
VERSION="35.20220424.3.0 (CoreOS)"
```



# Podman 4.1.0 Testing

Test on Podman 4.1.0 from latest on *stable* (there is a more recent image, **6.20220522.3.0** and I do not know why it is not pulling that instead):
```
$ brew list --versions qemu
qemu 7.0.0

$ podman --version
podman version 4.1.0

$ podman machine init \
    --memory 6144 \
    --cpus 4 \
    --disk-size 40 \
    --image-path stable
Downloading VM image: fedora-coreos-36.20220511.dev.0-qemu.aarch64.qcow2.xz: done  
Extracting compressed file
Image resized.

$ podman machine list
NAME                     VM TYPE     CREATED        LAST UP        CPUS        MEMORY      DISK SIZE
podman-machine-default*  qemu        7 minutes ago  7 minutes ago  4           6.442GB     42.95GB

$ podman machine start
Starting machine "podman-machine-default"
Waiting for VM ...
Error: dial unix /var/folders/77/qk6_53m11yb7t885qf6d32jw0000gn/T/podman/podman-machine-default_ready.sock: connect: connection refused
```

Now why it did not show the same error as in Podman 3.4.4 without changing the QEMU CPU and memory model? I have no clue.

Making the changes to **~/.config/containers/podman/machine/qemu/podman-machine-default.json** QEMU JSON configuration for CPU and memory model as suggested results in succes. Diff of the JSON file:
```
28c28
<   "cortex-a57",
---
>   "host",
30c30
<   "virt,highmem=off",
---
>   "virt,highmem=on",
51c51
< }
\ No newline at end of file
---
> }
```

Then starting the VM again results in:
```
$ podman machine start
Starting machine "podman-machine-default"
Waiting for VM ...
Mounting volume... /Users/cpolizzi:/Users/cpolizzi

This machine is currently configured in rootless mode. If your containers
require root permissions (e.g. ports < 1024), or if you run into compatibility
issues with non-podman clients, you can switch using the following command: 

	podman machine set --rootful

API forwarding listening on: /Users/cpolizzi/.local/share/containers/podman/machine/podman-machine-default/podman.sock

The system helper service is not installed; the default Docker API socket
address can't be used by podman. If you would like to install it run the
following commands:

	sudo /opt/homebrew/Cellar/podman/4.1.0/bin/podman-mac-helper install
	podman machine stop; podman machine start

You can still connect Docker API clients by setting DOCKER_HOST using the
following command in your terminal session:

	export DOCKER_HOST='unix:///Users/cpolizzi/.local/share/containers/podman/machine/podman-machine-default/podman.sock'

Machine "podman-machine-default" started successfully

$ podman version
Client:       Podman Engine
Version:      4.1.0
API Version:  4.1.0
Go Version:   go1.18.1
Built:        Thu May  5 15:07:47 2022
OS/Arch:      darwin/arm64

Server:       Podman Engine
Version:      4.1.0
API Version:  4.1.0
Go Version:   go1.18
Built:        Fri May  6 11:16:38 2022
OS/Arch:      linux/arm64

$ podman machine ssh cat /etc/os-release | grep ^VERSION=
VERSION="36.20220511.dev.0 (CoreOS)"
```

**Comment by shanesmith** [comment]
@cpolizzi interesting.... what's your hardware? I might have heard somewhere the crash only happens on M1 Pro.

I don't have a macOS 12.3 handy at the moment, could you try just setting `virt` (ie: without the `highmem`)?

**Comment by cpolizzi** [comment]
@shanesmith My hardware is an M1 Mac Mini with a mere 8GB of memory.

Total Cores:
```
$ sysctl -n hw.physicalcpu
8
```

Performance Cores:
```
sysctl -n hw.perflevel0.physicalcpu                   
4
```

Efficiency Cores:
```
sysctl -n hw.perflevel1.physicalcpu
4
```

It is seeming more like of an M1 Pro specific issue for some odd reason.

With just `virt` and no `himem`. Diff of **~/.config/containers/podman/machine/qemu/podman-machine-default.json**:
```
28c28
<   "cortex-a57",
---
>   "host",
30c30
<   "virt,highmem=off",
---
>   "virt",
51c51
< }
\ No newline at end of file
---
> }
```

The VM starts up just fine:
```
$ podman machine start
Starting machine "podman-machine-default"
Waiting for VM ...
Mounting volume... /Users/cpolizzi:/Users/cpolizzi

This machine is currently configured in rootless mode. If your containers
require root permissions (e.g. ports < 1024), or if you run into compatibility
issues with non-podman clients, you can switch using the following command: 

	podman machine set --rootful

API forwarding listening on: /Users/cpolizzi/.local/share/containers/podman/machine/podman-machine-default/podman.sock

The system helper service is not installed; the default Docker API socket
address can't be used by podman. If you would like to install it run the
following commands:

	sudo /opt/homebrew/Cellar/podman/4.1.0/bin/podman-mac-helper install
	podman machine stop; podman machine start

You can still connect Docker API clients by setting DOCKER_HOST using the
following command in your terminal session:

	export DOCKER_HOST='unix:///Users/cpolizzi/.local/share/containers/podman/machine/podman-machine-default/podman.sock'

Machine "podman-machine-default" started successfully
```

Listing verifies this:
```
$ podman machine list

podman machine list 
NAME                     VM TYPE     CREATED        LAST UP            CPUS        MEMORY      DISK SIZE
podman-machine-default*  qemu        5 minutes ago  Currently running  4           6.442GB     42.95GB
```

And further confirmed by connectivity:
```
$ podman machine ssh cat /etc/os-release | grep ^VERSION=
VERSION="36.20220511.dev.0 (CoreOS)"
```

**Comment by shanesmith** [comment]
It does not crash on my 13-inch 2020 MacBook Pro with Apple M1 chip (ie: non-Pro) with macOS 12.3.1. This indeed seems to be an issue specifically with macOS 12.3 and the Apple M1 Pro chip, which I believe is only in the 14-inch 2021 MacBook Pro at the moment.

**Comment by cpolizzi** [comment]
@shanesmith I have an individual from my customer whom has a 16-inch 2021 Apple M1 Max MacBook Pro on MacOS 12.3.1. He stated that once he made the change discussed here to the QEMU JSON configuration file and started the managed machine  caused his MacBook to crash. He said it happened twice.

**Comment by tideman** [comment]
For me the workaround by changing the memory size has no effect.
 I keep getting 
`Error: dial unix /var/folders/bh/q062sd9d5yvcjqnv/T/podman/podman-machine-default_ready.sock: connect: connection refused`
I also tried to downgrade qemu to 6.2.0
I am on a MacBook Pro (16-inch, 2021) Apple M1 PRO

**Comment by cpolizzi** [comment]
@tideman The CPU model has to be changed as well from `-cpu cortex-a57` to `-cpu host` as well as the memory model via the machine type. Machine type is specified via `-M` and is shorthand for `-machine`. The machine type remains as `virt` but the machine backend driver for memory access (`highmem=off`) needs to either be omitted entirely or specified as being enabled as `highmem=on`. `highmem=on` is the default.

Perhaps open a new GitHub issue so that you can provide your specific output and configuration?

**Comment by cpolizzi** [comment]
@nycnewman I agree with you that these seem to be two different issues. I have experienced this as @tideman mentioned, albeit extremely rarely. It is why I suggested creating a new issue for it. If I recall correctly I resolved it by wiping the temporary directory.  So in the output that @tideman shared I would have done:
```
sudo /var/folders/bh/q062sd9d5yvcjqnv/T/podman*
```
But, I could be wrong as well. One evening everything was working for this issue and the next day it no longer worked and I received the same results as we are now discussing. It is worth IMHO to open a new issue for this behavior and see what comes.

**Comment by wkoszek** [comment]
I'm experiencing what I believe is the same issue. I'm on:
- latest `brew upgrade` as of 2022-Jun-19
- MacOS Monterey 12.3.1 (21E258) with all updates except XCode
-  podman version 3.4.1

`podman machine init` never moves past `Waiting for VM` which I need to terminate with CTRL+C. Any workarounds are welcome.

Snippet below:

```
% podman machine start
INFO[0000] waiting for clients...
INFO[0000] listening tcp://0.0.0.0:7777
INFO[0000] new connection from  to /var/folders/c5/ctj1sbjn7gsc2qknql4csdb40000gn/T/podman/qemu_podman-machine-default.sock
Waiting for VM ...
^C
qemu-system-aarch64: terminating on signal 2 from pid 733 (<unknown process>)
% ERRO[0226] cannot receive packets from , disconnecting: cannot read size from socket: EOF
ERRO[0226] signal caught
```

**Comment by rhatdan** [comment]
Upgrade the host to 12.4.

**Comment by carlosgorges** [comment]
After some time debugging, I found the cause of this problem.

This problem is caused due to qemu 7.0.0 startup latency (3-5s) that occour in every first qemu execution after Mac Machine machine boots.

Podman has some bug that doesn't expect that the creation of socks files, done by the qemu call, can be delayed some seconds, and when podman tries to access the socks files, the qemu is not created them yet, showing the error "Error: dial unix <TEMPDIR>/podman/podman-machine-default_ready.sock: connect: connection refused".

To avoid this problem, just execute qemu, even with invalid options (just to initialize), before call "podman machine start".

                echo "* Podman VM machine for MACOSX is stoped, starting..."
                
                # workaround - initialize qemu before machine start to avoid socket error
                /usr/local/bin/qemu-system-x86_64 -machine q35,accel=hvf:tcg -cpu host -display none INVALID_OPTION >> /dev/null 2>&1

                podman machine start podman-machine-default
                ECODE=$?;if [ $ECODE -ne 0 ];then echo "* Error starting podman linux vm machine: $ECODE";exit $ECODE;fi

I hope help.
Carlos Eduardo Gorges.

**Comment by mixpix3ls** [comment]
> Please upgrade your podman to 4.1.1 and make sure your macos version is 12.4

👋🏽 @ashley-cui and all, thanks for looking into this. I'm able to start the machine (albeit after a long time given I used 2 vCPUs, 2GB of memory and 20GB disk specs for the VM).

**Comment by cpolizzi** [comment]
@mixpix3ls With your host upgraded to macOS 12.4 and using Podman 4.1.1 and ensuring that the linked fix #14563 is applied to your Podman configuration file for the QEMU `CmdLine` flags as described here and fixed in #14563 then you can run with more than your current VM sizing. Yesterday I confirmed with several folks that have M1 Pro Max SoC's that once they did these things they could size the VM to as the maximum number of performance cores available (8 in their case) and more than 3GB of memory without issue.

**Comment by baude** [comment]
we are cutting an 4.2 rc1 today.  Any chance you couild build that and try?

**Comment by waikontse** [comment]
Here is a snippet from my m1 pro on 4.2 rc1. I hope it is useful. Also, I've use podman machin set --rootful
before starting the machine. Removed the old machine with the old settings set up by podman 4.1.1
`
podman --version
podman version 4.2.0-rc1
➜  podman-4.2.0-rc1 podman machine start
Starting machine "podman-machine-default"
Waiting for VM ...
Error: qemu exited unexpectedly with exit code 1, stderr: qemu-system-aarch64: -fw_cfg name=opt/com.coreos/environment,string=bm9fcHJveHk9Ii5pbnRyYW5ldCwuaW5nLm5ldCwxMjcuMC4wLjEsbG9jYWxob3N0LCouZGV2LmF6dXJlLmNvbSx3ZWIud2hhdHNhcHAuY29tInxodHRwX3Byb3h5PSJodHRwOi8vbG9jYWxob3N0OjMxMjgifGh0dHBzX3Byb3h5PSJodHRwOi8vbG9jYWxob3N0OjMxMjgifA==: duplicate fw_cfg file name: opt/com.coreos/environment
`

**Comment by JonasJonny** [comment]
Today I had:
`podman/podman-machine-default_ready.sock: connect: connection refused`

What worked for me:

```
1. podman machine stop
2. podman machine rm
3. podman machine init
4. podman machine start
```

**Comment by ericcurtin** [comment]
> My workaround is to delete the VM and start from scratch
> 
> ```
> $ podman machine stop
> $ podman machine rm --force
> $ podman machine init --cpus 2 --memory 2048 --disk-size 20
> $ podman machine start
> ```
> 
> version
> 
> ```
> $ podman --version
> podman version 4.1.0
> $ qemu-system-aarch64 --version 
> QEMU emulator version 7.0.0
> ```

Highly recommend this solution if you are happy to throw away everything by the way...

**Comment by mkenigs** [comment]
> I still have problems getting this to work. First, since I am on Podman v4.1.1 should not `-cpu` be `host` and `highmem=on` be that by default now? I still have to make that change manually here. And also after doing that I am unable to get it running.
> 
> I am using nix as package manager if that is of any help

Also using nix, and for some reason I was still running into this until 4.2

**Comment by paulatwork** [comment]
EXCELLENT fix from Carlos Eduardo Gorges, on stackoverflow.
Caused due to qemu 7.0.0 startup latency (3-5s).
To fix .. execute qemu, to initialize, before call "podman machine start".
Works for me .. 
https://stackoverflow.com/questions/72404926/cant-get-podman-to-run-on-an-mac-m1-monterey

**Comment by justkash** [comment]
This is a problem with virtualizing memory in newer versions of Qemu (7.0+) on aarch64 machines. I resolved the issue by updating the `~/.config/containers/podman/machine/qemu/podman-machine-default.json` file before starting the podman machine VM. Here are the steps that worked for me:
1. Create machine VM with `podman machine init` ( with `-v $HOME:$HOME` if you want to be able to mount volumes)
2. Update `-M` entry to `virt,highmem=off` in `~/.config/containers/podman/machine/qemu/podman-machine-default.json`
3. Save file and run `podman machine start`
